package norns.rpc;

message Request {

    // oneof would be better here, but the version of protoc provided by
    // CentOS 7 does not support it yet
    enum Type {
        SUBMIT_IOTASK = 1;
        REGISTER_JOB = 2;
        UPDATE_JOB = 3;
        UNREGISTER_JOB = 4;
        ADD_PROCESS = 5;
        REMOVE_PROCESS = 6;
    }

    required Type type = 1;

    // I/O task descriptor
    message Task {
    }
    optional Task task = 2;

    // job descriptor
    message Job {
        repeated string hosts = 1;

        // backend descriptor
        message Backend {
            required int32 type = 1;
            required string mount = 2;
            required int32 quota = 3;
        }

        repeated Backend backends = 3;
    }

    // process descriptor
    message Process {
        required int64 pid = 1;
        required int64 gid = 2;
    }

    optional uint32 jobid = 3;
    optional Job job = 4;
    optional Process process = 5;
}

message Response {
    // most responses only need to return a status code
    required uint32 status = 2;
}


